### PR TITLE
[#9557] Standardise the way of deleting entities in storage component

### LINKED ISSUE #9557: Standardise the way of deleting entities in storage component
The deletion APIs are non-standard and have the following problems.

> Unnecessary deletion query.

The current deletion method need to query DB twice when deleting (some) entities. This is not necessary. When keys are available, we can issue deletion request directly -> so only 1 query.

> Telescoping deletion methods.

Take response comment entities as an example. The app can delete comments of a response, question, session, course. There are four methods here. For response, the app can delete responses of a question, session and course. There are three methods here. There are 2 methods and 1 method for question and session respectively. Anyway, there is a hierarchy structure here for comment -> response -> question -> session -> course. We can use builder pattern to solve this.

> Slow one-by-one cascade deletion

For some reasons, comment entity stores information of a course, session and question. In database normalisation, this attributes are redundant because once its response is identified, its corresponding question, session and course are automatically identified. Same reasoning for response and question. 

Instead of removing the attributes, we can make use of it for faster deletion. For example, if we want to deletion all responses of a course (session or question), we can issue just one batch deletion query.

> Bugs in cascade logic
- Account deletion will result in orphan course
- Instructors' responses are not deleted when an instructor is deleted
- Respondent's lists are not updated properly when doing cascade deletion

### PR DESCRIPTION
**Pull request by xpdavid**
Fixes #9557 

**Outline of Solution**

- Issue deletion query directly when key is available
- Introduce `AttributesDeletionQuery` to solve the problem of telescoping deletion methods
- Make use of `AttributesDeletionQuery` and delete entities by batch for cascade deletion
    - For cascade deletion involving responses, there is no way to do batch deletion (we have to do it one by one) because we need update the respondents lists.
- Write full documentations for each deletion method
- Full tests are written for each deletion method

### CODE REVIEW

**Comment by wkurniawan07** [comment]
@xpdavid can resolve the merge conflicts?

File: docs/design.md [hunk 1]
```diff
@@ -191,10 +185,7 @@ Represented by the `*Db` classes. These classes act as the bridge to the GAE Dat
...
 Note: 'Eventual consistency' here means it takes some time for a database operation to propagate across all serves of the Google's distributed datastore. As a result, the data may be in an inconsistent states for short periods of time although things should become consistent 'eventually'. For example, an object we deleted may appear to still exist for a short while.
 
-Implementation of Transaction Control has been decided against due to limitations of GAE environment and the nature of our data schema. Note that this decision was taken before 2014 and if the circumstances leading to that decision have changed, it may be reconsidered.
-
-General:
-+ If `Null` is passed as a parameter, the corresponding value is **NOT** modified, as per the KeepExistingPolicy that was previously mentioned.
+Implementation of Transaction Control has been minimized due to limitations of GAE environment and the nature of our data schema.
```
[the remark this answers is not in the document]
Comment (xpdavid) [reply docs/design.md]: Used for the update of feedback session respondent lists.

BTW, I need some time to resolve the conflicts introduced by another another refactor PR :P

**Comment by xpdavid** [comment]
@wkurniawan07 Conflicts resolved.

File: src/main/java/teammates/logic/core/CoursesLogic.java [hunk 2]
```diff
@@ -646,31 +650,26 @@ public CourseAttributes updateCourseCascade(CourseAttributes.UpdateOptions updat
...
-        Assumption.assertNotNull("Supplied parameter was null", instructorList);
+        if (getCourse(courseId) == null) {
+            return;
+        }
 
-        List<String> softDeletedCourseIdList = instructorList.stream()
-                .filter(instructor -> coursesDb.getCourse(instructor.courseId).isCourseDeleted())
-                .map(InstructorAttributes::getCourseId)
-                .collect(Collectors.toList());
+        coursesDb.deleteCourse(courseId);
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/CoursesLogic.java]: Should this be done last? That way, if any of the following operations fail, you can safely re-run this same delete operation. If this is done first, if any of the following operations fail, you'll end up with orphan entities.

File: src/main/java/teammates/logic/core/AccountsLogic.java [hunk 3]
```diff
@@ -252,16 +252,29 @@ public void makeAccountInstructor(String googleId) throws InvalidParametersExcep
...
-        instructorsLogic.deleteInstructorsForGoogleIdAndCascade(googleId);
-        studentsLogic.deleteStudentsForGoogleIdAndCascade(googleId);
+
+        // to prevent orphan course
+        List<InstructorAttributes> instructorsToDelete =
+                instructorsLogic.getInstructorsForGoogleId(googleId, false);
+        for (InstructorAttributes instructorToDelete : instructorsToDelete) {
+            if (instructorsLogic.getInstructorsForCourse(instructorToDelete.getCourseId()).size() <= 1) {
+                // the instructor is the last instructor in the course
+                coursesLogic.deleteCourseCascade(instructorToDelete.getCourseId());
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/AccountsLogic.java]: Similarly here, should the instructor be deleted first before the course?

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 4]
```diff
@@ -1373,53 +1368,25 @@ public void unpublishFeedbackSession(FeedbackSessionAttributes sessionToUnpublis
...
+     * Deletes a feedback session cascade to its associated questions, responses and comments.
      */
     public void deleteFeedbackSessionCascade(String feedbackSessionName, String courseId) {
+        AttributesDeletionQuery query = AttributesDeletionQuery.builder()
+                .withCourseId(courseId)
+                .withFeedbackSessionName(feedbackSessionName)
+                .build();
+        fqLogic.deleteFeedbackQuestions(query);
+        frLogic.deleteFeedbackResponses(query);
+        frcLogic.deleteFeedbackResponseComments(query);
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: Similarly here; response comments should be deleted before the responses, which should be deleted before the questions?

File: src/main/java/teammates/logic/core/StudentsLogic.java [hunk 5]
```diff
@@ -382,29 +383,35 @@ private String getTeamInvalidityInfo(List<StudentAttributes> mergedList) {
...
-        for (StudentAttributes student : students) {
-            fsLogic.deleteStudentFromRespondentsList(student);
+        StudentAttributes student = getStudentForEmail(courseId, studentEmail);
+        if (student == null) {
+            return;
         }
-        studentsDb.deleteStudentsForGoogleId(googleId);
+
+        studentsDb.deleteStudent(courseId, studentEmail);
+        frLogic.deleteFeedbackResponsesInvolvedStudentOfCourseCascade(courseId, studentEmail, student.getTeam());
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/StudentsLogic.java]: Similarly here?

**Comment by xpdavid** [comment]
Yup. It would be better to delete the `master` entity at the end.

**Comment by damithc** [comment]
go ahead and merge 👍

**Pull request merged**